src/item.h

```diff
@@ -1194,25 +1194,47 @@ class item : public JsonSerializer, public JsonDeserializer, public visitable<it
         item * gunmod_find( const itype_id& mod );
         const item * gunmod_find( const itype_id& mod ) const;
 
-        /**
-         * Returns currently active auxiliary (@ref is_auxiliary_gunmod) gun mod item.
-         * May return null if there is no such gun mod or if the gun is not in the
-         * auxiliary mode (@ref is_in_auxiliary_mode).
-         */
-        item * gunmod_current();
-        item const * gunmod_current() const;
-
         /*
          * Checks if mod can be applied to this item considering any current state (jammed, loaded etc.)
          * @param alert whether to display message describing reason for any incompatibility
          * @param effects whether temporary efects (jammed, loaded etc) are considered when checking
          */
         bool gunmod_compatible( const item& mod, bool alert = true, bool effects = true ) const;
 
-        /**
-         * Burst size (see ranged.cpp), includes effects from installed gunmods.
-         */
-        int burst_size() const;
+        struct gun_mode {
+            std::string mode; /** name of this mode */
+            item *target;     /** pointer to base gun or attached gunmod */
+            int qty;          /** burst size or melee range */
+            bool melee;       /** if true perform a melee attach as opposed to shooting */
+
+            operator bool() const { return target != nullptr; }
+
+            item &operator*() { return *target; }
+            const item &operator*() const { return *target; }
+
+            item *operator->() { return target; }
+            const item *operator->() const { return target; }
+        };
+
+        /** Get all possible modes for this gun inclusive of any attached gunmods */
+        std::map<std::string, const item::gun_mode> gun_all_modes() const;
+
+        /** Check if gun supports a specific mode returning an invalid/empty mode if not */
+        const gun_mode gun_has_mode( const std::string& mode ) const;
```

Comment:
I'd say it would be obvious enough from the fact that `gun_get_mode` has an argument.

Even if `gun_get_mode` had an overload with no argument (returning current mode), I'd say it would be pretty obvious from context what is going on.